src/nupic/regions/BacktrackingTMRegion.hpp

```diff
@@ -0,0 +1,157 @@
+/* ---------------------------------------------------------------------
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2018, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ *
+ * Author: David Keeney, June, 2018  (ported from Python)
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+ * Declarations for TMRegion class
+ */
+
+//----------------------------------------------------------------------
+
+#ifndef NTA_BACKTRACKING_TMREGION_HPP
+#define NTA_BACKTRACKING_TMREGION_HPP
+
+#include <memory>
+#include <nupic/engine/RegionImpl.hpp>
+#include <nupic/engine/Spec.hpp>
+#include <nupic/ntypes/Value.hpp>
+#include <nupic/algorithms/BacktrackingTMCpp.hpp>
```

Comment:
nit, consider `git mv` to BacktrackingTM.hpp (drop the Cpp from name)

Reply:
If I rename both the algorithm and the RegionImpl, how about something shorter like
```
  BackTM.hpp/cpp   and BackTMRegion.hpp/cpp
```
Would that confuse people too much?

Reply:
There's no cpp version, so I guess you've free hands with the name

Reply:
Lets not rock the boat too much...I will rename the algorithm BacktrackingTM.